Add in-line emoji styling to match reaction styling

blenderskool opened the pull request:
Emojis in messages are styled to match the styling of reactions. Reactions are additionally made bold to better differentiate between them.

Before:
![before](https://user-images.githubusercontent.com/21107799/66712584-3c21e680-edbc-11e9-84df-adb527e2ea63.png)

After:
![after](https://user-images.githubusercontent.com/21107799/66712597-77bcb080-edbc-11e9-9f68-2756a1a37be1.png)

blenderskool commented:
Fixes #412

sumanthvrao commented:
@blenderskool you can run `pytest && ./tools/run-mypy` to run the tests locally. Are you familiar with pytest?  You can look at existing tests and see how you can fix the errors and add new tests. 

The diff in your commits shows that the second commit only has unnecessary spaces. You could drop the commit (`git rebase` can help you here)

Feel free to reach out incase you have further doubts.

blenderskool commented:
@sumanthvrao Ok I'll make the necessary changes as you mentioned 🙂

neiljp commented:
@blenderskool Thanks for working on this ;)

Did you test the extended colo[u]rs you added to the gruvbox scheme? Where did they come from?

You might like to use `gitlint` as described in the README, and also see how we generally style commit text based on looking at `git log` (or using `tig`).

While #412 (based on the original discussion) suggests using an entirely different style for reactions - and we can continue to do this - I'm wondering if we should reserve a 'bold' styling of whatever the base reactions colo[u]r is for when a user has contributed to that reaction @sumanthvrao @amanagr Thoughts?

neiljp commented:
@blenderskool Oh, and please pop in on the zulip stream at chat.zulip.org if you want to discuss any aspect of development :)

blenderskool commented:
@neiljp @sumanthvrao I updated the tests to match the added `emoji` styling. Also fixed the gruvbox color that I had added to a `neutral_purple` color I found here https://github.com/morhetz/gruvbox/blob/master/colors/gruvbox.vim

neiljp commented:
@blenderskool Thanks for the update - I'm glad to be able to get back to this! This looks *almost* ready to go, but I've described some aspects of how you seem to have structured the commits below. I could change them myself, but hopefully you might find the following useful to work through, though please ask questions if you don't follow or agree with my reasoning :)

Overall your branch passes the tests, but individually the commits don't. It's often useful to have lots of small 'working' commits during development (which perhaps is what you have here), but prior to merging (or a near-final PR) it's useful to have commits which individually don't break anything and are cleaner. In other projects, commits can get squashed together when a branch is merged, which hides this. However we rebase on merge here, and cleaner commits can make it easier to review code.

This behavior (not passing tests from one commit to another) can also suggest that elements in each commit aren't grouped quite right between commits. Could you comment on the following:
* Is the last commit a fix to the style in the first one? If so, since this is new code you're adding, do those two commits belong squashed together?
* Similarly, is the second commit a fix to the first?

Depending on how those commits get squashed together (or not), you can determine where the tests changes should go, so as to not break the fact that the tests pass from one commit to another :)

blenderskool commented:
@neiljp Thanks for the comments. I agree with what you've mentioned about the commit structure. I'll be more thoughtful regarding this in the future.
How do I go about fixing this with the commits I made in this branch?
- The last commit is indeed a fix for gruvbox style in the first commit.
- The second commit adds custom icon support to the first commit where I didn't notice that it was being handled separately.

neiljp commented:
@blenderskool Depending upon how big the changes are, you might find an **interactive rebase** useful, which allows you to reorder and merge (squash, or fixup - when you're confident!) commits. Sometimes this can lead to conflicts - but you can always escape with a `git rebase --abort` if things get confusing :)

If the later commits are fixes to the first commit - not just developments on top of it, which is a subtle difference - then particularly if they're small then they could just be all squashed together into one commit, with the commit text updated.

There are many git resources out there, but you may find the Zulip git guide useful, specifically this section at the moment: https://zulip.readthedocs.io/en/latest/git/fixing-commits.html

blenderskool commented:
@neiljp I squashed and reordered some commits which made logical sense. Thanks for the advice.

neiljp commented:
@blenderskool Thanks for working on this - I hope this was useful for you too. The code looks a lot simpler to understand when looking at just these commits :+1: I checked manually and things still look to be working fine overall :)

I ran `tools/lint-all && pytest` on your branch and after checking out `HEAD~`. The branch passes, as per CI results, but the latter intermediate checkout doesn't. This is since the tests are added prior to the code that implements them. Adding them afterwards instead would be OK, but generally we include the tests in the same commit as the code that would otherwise break them.

Given the above, but that the code is essentially fine, I've squashed the two commits and merged them manually as 9a708277d67d7285066e64d09e49da784fa197c2. Other than making a minor change to the commit text for accuracy (match -> similar to), I also adjusted the background color of the blue & light themes for the added style.
